### PR TITLE
Add Materialized View support to the Iceberg Glue Catalog

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

Adds support for Materialized Views when using the Iceberg connector with Glue.

> Is this change a fix, improvement, new feature, refactoring, or other?

New feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Connector, Iceberg.

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

Materialized Views are already described in the Iceberg documentation.

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Iceberg
* Materialized View support to the Iceberg Glue Catalog
```

### CODE REVIEW

**Comment by findepi** [comment]
@alexjo2144 please rebase your PR (i guess it conflicted with https://github.com/trinodb/trino/pull/11766)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 1]
```diff
@@ -553,30 +574,162 @@ public void dropView(ConnectorSession session, SchemaTableName schemaViewName)
...
     {
-        return Optional.empty();
+        Optional<com.amazonaws.services.glue.model.Table> maybeTable = getTable(schemaViewName);
+        if (maybeTable.isEmpty()) {
+            return Optional.empty();
+        }
+
+        com.amazonaws.services.glue.model.Table table = maybeTable.get();
+        Map<String, String> tableParameters = table.getParameters();
+        if (!isPrestoView(tableParameters) || !isHiveOrPrestoView(table.getTableType()) || !tableParameters.containsKey(STORAGE_TABLE)) {
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:672]: Why don't we use `!(isPrestoView(table.getParameters()) && table.getParameters().get(TABLE_COMMENT).equalsIgnoreCase(ICEBERG_MATERIALIZED_VIEW_COMMENT))` like in `listMaterializedViews` ?
Maybe extract that into a `isTrinoMaterializedView` function

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergGlueCatalogConnectorSmokeTest.java [hunk 2]
```diff
@@ -121,14 +121,6 @@ public void testShowCreateTable()
                         schemaPath()));
     }
 
-    @Test
-    @Override
-    public void testMaterializedView()
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergGlueCatalogConnectorSmokeTest.java:126]: Smoke tests are testing only cases around CREATE, SELECT, SHOW CREATE and DROP.
We should be testing CREATE OR REPLACE, listing of MVs and RENAME as well with Glue. Is that already covered by other tests ?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java (hunk 1 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java]: Added `isTrinoMaterializedView`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergGlueCatalogConnectorSmokeTest.java (hunk 2 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergGlueCatalogConnectorSmokeTest.java]: I didn't add any. Do you know if there's coverage in BaseConnectorTest? Yuya has a PR running that against Glue https://github.com/trinodb/trino/pull/11746

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 3]
```diff
@@ -553,30 +573,161 @@ public void dropView(ConnectorSession session, SchemaTableName schemaViewName)
...
+                return;
+            }
+            throw new TrinoException(ALREADY_EXISTS, "Materialized view already exists: " + viewName);
+        }
+
+        if (existing.isPresent()) {
+            // drop the current storage table
+            String oldStorageTable = existing.get().getParameters().get(STORAGE_TABLE);
+            if (oldStorageTable != null) {
+                dropTable(session, new SchemaTableName(existing.get().getDatabaseName(), oldStorageTable));
```
Comment (raunaqmorarka) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:626]: We should be consistent about whether we wrap it in try/catch with `dropMaterializedView` here. Might be worth extracting a common method for dropping storage table and deleting MV definition.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/ViewReaderUtil.java [hunk 4]
```diff
@@ -145,6 +147,11 @@ public static boolean isHiveOrPrestoView(String tableType)
         return tableType.equals(TableType.VIRTUAL_VIEW.name());
     }
 
+    public static boolean isTrinoMaterializedView(String tableType, Map<String, String> tableParameters)
+    {
+        return isHiveOrPrestoView(tableType) && isPrestoView(tableParameters) && tableParameters.containsKey(STORAGE_TABLE);
```
Comment (raunaqmorarka) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/ViewReaderUtil.java:152]: I think `tableParameters.containsKey(STORAGE_TABLE)` should be replaced by `table.getParameters().get(TABLE_COMMENT).equalsIgnoreCase(ICEBERG_MATERIALIZED_VIEW_COMMENT)` here. Then any code using storage table from the table parameters should `verify(tableParameters.containsKey(STORAGE_TABLE))`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergGlueCatalogConnectorSmokeTest.java (hunk 2 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergGlueCatalogConnectorSmokeTest.java]: For now I converted `TestIcebergMaterializedViews` to `BaseIcebergMaterializedViewTest` and added both a Hive and Glue backed implementation.

**Review by raunaqmorarka — APPROVED** [review]

**Comment by findepi** [comment]
@alexjo2144 please squash and rebase

**Comment by alexjo2144** [comment]
Squashed and rebased

**Comment by findepi** [comment]
> @findepi can you create a Trino mirror branch to run the Glue tests on?

https://github.com/trinodb/trino/pull/11833

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 5]
```diff
@@ -556,31 +580,162 @@ public void dropView(ConnectorSession session, SchemaTableName schemaViewName)
...
+        String storageTableName = view.getParameters().get(STORAGE_TABLE);
+        if (storageTableName != null) {
+            try {
+                dropTable(session, new SchemaTableName(schemaName, storageTableName));
+            }
+            catch (TrinoException e) {
+                LOG.warn(e, "Failed to drop storage table '%s' for materialized view '%s'", storageTableName, viewName);
+            }
+        }
+        deleteTable(schemaName, viewName);
```
Comment (homar) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:664]: Silly question, does it make sense to delete table that was not dropped correctly?

**Review by homar — APPROVED** [review]

**Comment by alexjo2144** [comment]
Updated the `GET_TABLE` counts in `TestIcebergGlueCatalogAccessOperations`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 6]
```diff
@@ -560,31 +580,162 @@ public void dropView(ConnectorSession session, SchemaTableName schemaViewName)
...
+            boolean replace,
+            boolean ignoreExisting)
     {
-        throw new TrinoException(NOT_SUPPORTED, "createMaterializedView is not supported for Iceberg Glue catalogs");
+        Optional<com.amazonaws.services.glue.model.Table> existing = getTable(viewName);
+
+        // It's a create command where the materialized view already exists and 'if not exists' clause is not specified
+        if (!replace && existing.isPresent()) {
+            if (ignoreExisting) {
+                return;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:624]: This should succeed only if this is an MV

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 7]
```diff
@@ -560,31 +580,162 @@ public void dropView(ConnectorSession session, SchemaTableName schemaViewName)
...
+            }
+            throw new TrinoException(ALREADY_EXISTS, "Materialized view already exists: " + viewName);
+        }
+
+        existing.ifPresent(table -> dropMaterializedView(session, table));
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:629]: i'd would be better if CREATE OR REPLACE was atomic. Can we do this?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 8]
```diff
@@ -560,31 +580,162 @@ public void dropView(ConnectorSession session, SchemaTableName schemaViewName)
...
     }
 
     @Override
     public void dropMaterializedView(ConnectorSession session, SchemaTableName viewName)
     {
-        throw new TrinoException(NOT_SUPPORTED, "dropMaterializedView is not supported for Iceberg Glue catalogs");
+        com.amazonaws.services.glue.model.Table view = getTable(viewName)
+                .orElseThrow(() -> new MaterializedViewNotFoundException(viewName));
+
+        dropMaterializedView(session, view);
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:648]: this should succeed only if this is an MV. Currently it seems to drop any table

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java [hunk 9]
```diff
@@ -560,31 +580,162 @@ public void dropView(ConnectorSession session, SchemaTableName schemaViewName)
...
     @Override
     public void renameMaterializedView(ConnectorSession session, SchemaTableName source, SchemaTableName target)
     {
-        throw new TrinoException(NOT_SUPPORTED, "renameMaterializedView is not supported for Iceberg Glue catalogs");
+        boolean newTableCreated = false;
+        try {
+            Optional<com.amazonaws.services.glue.model.Table> table = getTable(source);
+            if (table.isEmpty()) {
+                throw new TableNotFoundException(source);
+            }
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java:717]: validate it's MV before proceeding

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java [hunk 10]
```diff
@@ -176,7 +176,7 @@ public void testSelectWithFilter()
 
             assertGlueMetastoreApiInvocations("SELECT * FROM test_select_from_where WHERE age = 2",
                     ImmutableMultiset.builder()
-                            .addCopies(GET_TABLE, 2)
+                            .addCopies(GET_TABLE, 3)
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:179]: :sadpanda:

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.plugin.hive.metastore.glue.AwsSdkUtil.getPaginatedResults;
+import static io.trino.testing.sql.TestTable.randomTableSuffix;
+
+public class TestIcebergGlueCatalogMaterializedViewTest
+        extends BaseIcebergMaterializedViewTest
+{
+    String schemaName = "iceberg_materialized_view_test_" + randomTableSuffix();
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java:44]: Will something eventually clean these up when test is aborted in the middle (`cleanup` isn't invoked or fails)?

Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java:44]: `private final`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        File tmp = Files.createTempDirectory("test_iceberg").toFile();
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java:50]: `tempDir` 

delete contents (`closeAfterClass`?)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java (hunk 7 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java]: Yep, switched to use `updateTable`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java (hunk 10 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogAccessOperations.java:179]: Agreed

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java (hunk 12 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java]: Added a `deleteOnExit`

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java (hunk 11 above)
Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java]: I was going to put up a PR for cleanup across all these tests. I was thinking of adding a `iceberg_integration_test_schema` flag or something like that in the Glue schema properties to be able to filter them easily

Comment (alexjo2144) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java]: https://github.com/trinodb/trino/pull/11903

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static io.trino.plugin.hive.metastore.glue.AwsSdkUtil.getPaginatedResults;
+import static io.trino.testing.sql.TestTable.randomTableSuffix;
+
+public class TestIcebergGlueCatalogMaterializedViewTest
+        extends BaseIcebergMaterializedViewTest
+{
+    private final String schemaName = "iceberg_materialized_view_test_" + randomTableSuffix();
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/catalog/glue/TestIcebergGlueCatalogMaterializedViewTest.java:44]: let's start it with test_ (https://github.com/trinodb/trino/pull/11903#discussion_r849174508)

**Comment by findepi** [comment]
( squashed and rebased )

**Comment by findepi** [comment]
( applied https://github.com/trinodb/trino/pull/11780/#discussion_r849275169 )

**Comment by findepi** [comment]
Updated https://github.com/trinodb/trino/pull/11833

**Review by findepi — APPROVED** [review]

**Pull request merged**